initial commit (#1)
mosabami/cost-optimized-workloads-aks: cost-optimization-aks-content.md

In this document, we'll discuss features and options, and the best practices for running and managing Kubernetes workloads on AKS in a cost optimized way to take advantage of the scalability provided by Azure. 

> [!NOTE]
> This document assumes you are familiar with Kubernetes, Azure, AKS and autoscaling. If you need a refresher, see the [**Related articles**](#related-articles) section for resources.

## Introduction

> As Kubernetes gains widespread adoption, a growing number of enterprises and platform-as-a-service (PaaS) and software-as-a-service (SaaS) providers are using multi-tenant Kubernetes clusters for their workloads. This means that a single cluster might be running applications that belong to different teams, departments, customers, or environments. The multi-tenancy provided by Kubernetes lets companies manage a few large clusters, instead of multiple smaller ones, with benefits such as appropriate resource utilization, simplified management control, and reduced fragmentation.

Reviewer: Couple small suggestions.  Kubernetes really doesn't have multi-tenancy "provided" -- you kinda need to stitch that together.

```suggestion
As Kubernetes gains widespread adoption, a growing number of enterprises and platform-as-a-service (PaaS) and software-as-a-service (SaaS) providers are using multi-tenant Kubernetes clusters for their workloads. This means that a single cluster is running applications that belong to different teams, departments, customers, or environments. This lets companies manage a few large clusters, instead of multiple smaller ones, with benefits such as appropriate resource utilization, simplified management control, and reduced fragmentation.
```

> Over time, some of these companies with fast-growing Kubernetes clusters start to experience a disproportionate increase in cost. This happens because traditional companies that embrace cloud-based solutions like Kubernetes don't have developers and operators with cloud expertise. This lack of cloud readiness leads to applications becoming unstable during autoscaling (for example, traffic volatility during a regular period of the day), sudden bursts, or spikes (such as TV commercials or peak scale events like Black Friday and Cyber Monday). In an attempt to "fix" the problem, these companies tend to over-provision their clusters the way they used to in a non-elastic environment. Over-provisioning results in considerably higher CPU and memory allocation than what applications use for most of the day.

Reviewer: > This happens because traditional companies that embrace cloud-based solutions like Kubernetes don't have developers and operators with cloud expertise.

I'd try to find another phase here.  This is pretty harsh to say.  You can be a company that experiences this problem even if you have cloud expertise.  And cloud expertise isn't an "on" or "off" state, so hard to say when you have it or not.  The premise of "over-provision to fix scaling issues" is a good angle to lead with in this article, I'd suggest doing it though with less direct "attack" on the customer's capabilities, experiences, or staff.

This document provides best practices and highlights some of the features that help with cost-optimization of Kubernetes workloads on AKS. The table below outlines this approach.

The foundation of building cost-optimized applications is spreading the cost-saving culture across teams. Beyond moving cost discussions to the beginning of the development process, this approach forces you to better understand the environment that your applications are running in—in this context, the AKS environment.

> In order to achieve low cost and application stability, you must correctly set or tune some features and configurations (such as autoscaling, machine types, and region selection). Another important consideration is your workload type because, depending on the workload type and your application's requirements, you must apply different configurations in order to further lower your costs. Finally, you must monitor your spending and create guardrails so that you can enforce best practices early in your development cycle.

Reviewer: remember to keep the story focused on multi-tenant like you led with -- so in this example moved to plural.

```suggestion
In order to achieve low cost and application stability, you must correctly set or tune some features and configurations (such as autoscaling, machine types, and region selection). Another important consideration is your workloads' types and requirements, because you must apply different configurations in order to further lower your costs. Finally, you must monitor your spending and create guardrails so that you can enforce best practices early in your development cycle.
```

    :::column-end:::
   :::column:::

      #### Build a cost-saving Culture
      Better align developers with your business by designing and iterating for cost optimization early in the development cycle
      [Spread the cost saving culture](###spread-the-cost-saving-culture)\
    :::column-end:::  
:::row-end:::

:::row:::
    :::column:::

      #### Monitor your spending
      Monitor your environment and enforce cost-optimized practices
      [Observe your clusters and watch for recommendations](###observe-your-clusters-and-watch-for-recommendations)\
      [Monitor costs breakdown using Kubecost](###monitor-costs-breakdown-using-Kubecost)\
      [Understand how Metrics Server works and monitor it](###understand-how-Metrics-Server-works-and-monitor-it)\
      [Use Kubernetes Resource Quotas](###use-kubernetes-resource-quotas)\
      [Enforce cost-optimized practices with Azure Policy](###enforce-cost-optimized-practices-with-azure-policy)\
      [Design your CI/CD pipeline to enforce cost-saving practices]
    :::column-end:::
:::row-end:::

## Optimize your AKS Environment
Optimize cost with autoscaling, sizing, machine type, etc
### Optimize AKS Autoscaling

> ### Choose the right machine type

Reviewer: ```suggestion
### Choose the right machine type(s)
```

> ### Select the appropriate region

Reviewer: ```suggestion
### Select the appropriate region(s)
```

### Sign up for Reserved Instances
### Review small development clusters
### Review your monitoring strategies
### Review inter-region egress traffic

> ## Build Cloud-native applications

Reviewer: The article leds with multi-tenant.  I think it's perfectly fair to dive down to the responsibilities of a single tenant like you have here.  Because it's the union of all the tenants that cause the user to arrive at the problem statement at the start of this.  Remind the users of this here.

Prepare your applications to run in such an environment
### Understand your application capacity
### Ensure your application can grow

> ### Use appropriate requests and limits

Reviewer: Might be good to mention "noisy neighbor" control in this section.  Every app needs to be harmonious and follow the rules.  No bounded growth, etc.

### Ensure your container is as lean as possible
### Add pod disruption budget to your application

> ### Set meaningful readiness and liveliness probes

Reviewer: We often talk about applications needing "meaningful" scale metrics (such as queue depth), however architecture is about tradeoffs.  Consider the following.  If you a flappy queue -- it gets filled up FAST and in a large burst.  If you used KEDA based on queue depth, you could easily spin up 10s of VMs to drain that queue super quick, and based on your autoscaller settings, have those VMs spin down really quick. -- Gotcha -- we charge per hour.  You are paying WAY more by mass scale out and then back in to drain the queue quickly than if you were to drain it slower with less overall compute.  I think cost optimization on AKS really is mostly about tuning workloads to scale no more than is necessary, know your scale patterns - if you scale out shortly after you scale in, repeatedly. -- then it makes the most sense to have a longer tail on your deallocation (scale-in) configuration. 

I think the heart of this article really is the dance that goes between application scalability (just the right amount) and controlling the lifetimes of the underlying VM resources.  The more you can talk about VMs and their pricing models and then tie that back into the cluster autoscaller tunability and then turn that back into HPA and KEDA... that'll make this article pop.

> ### Ensure your container is shutting down gracefully

Reviewer: > Ensure your container is shutting down gracefully

And starting.  This goes with the readiness/liveliness probes, but I've seen a poorly written application that would go into crash restart hell, and based on HPA, cause massive scale out attempts on the cluster -- all because the order in which the workload was deployed was wrong (dependency deployed AFTER the thing that depends on it)  That can cause a cluster to temporarily balloon, which adds hourly cost quickly because it'll settle right back down later.

> ### Consider using retries with exponential backoff

Reviewer: > Consider using retries with exponential backoff

Not sure how this will tie into cost in any significant way, but eager to learn.  I haven't really thought about the backoff math when it comes to cost -- how exponential is more cost effective than another type of backoff.  Interesting.

## Consider Workload Types
Optimize cost by configuring your environment according to your workload type
### Prepare your environment to fit your workload type

## Build a cost-saving Culture
Better align developers with your business by designing and iterating for cost optimization early in the development cycle
Spread the cost saving culture

> ### Spread the cost saving culture

Reviewer: I love this section.

Reviewer: Applications are only as good as they are written.  If you write a memory hungry, CPU hungry application, it's going to COST more to run.  Period.  If you can save 5% CPU or MEMORY usage with optimized code, you've just saved your comapny 5% on their spend. That's the mindset.  a 5% reduced resource footprint = 5% less compute footprint.  And when you have 100s of nodes, that 5% is massive $.

Reviewer: This goes into your CI/CD pipeline talk as well... if you have performance metrics from prior deployments, you'll know quickly if you're going to have to "spend more" on this or "spend less" on the new deployment based on the performance profiling done during the deployment.  Eager to hear about that in this story!  Especially how AKS can be measured like that.
